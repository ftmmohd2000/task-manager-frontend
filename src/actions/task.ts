import TaskType from "../types/task";
import client from "../rest/api";

export interface InitAction {
  type: "@@INIT";
}

export interface SetTasksAction {
  type: "SET_TASKS";
  tasks: TaskType[];
}

export interface AddTaskAction {
  type: "ADD_TASK";
  task: TaskType;
}

export interface RemoveTaskAction {
  type: "REMOVE_TASK";
  id: string;
}

export interface EditTaskAction {
  type: "EDIT_TASK";
  id: string;
  updates: Partial<TaskType>;
}

export interface ClearTasksAction {
  type: "CLEAR_TASKS";
}

export const startSetTasks = () => {
  return (dispatch: any, getState: any) => {
    return client
      .get("/tasks", {
        headers: {
          Authorization:
            "Bearer " + getState().auth.user.token
        }
      })
      .then((res) => {
        dispatch(setTasks(res.data));
      })
      .catch(() => {});
  };
};

export const startAddTask = (
  description: string,
  completed: boolean
) => {
  return (dispatch: any, getState: any) => {
    return client
      .post(
        "/tasks",
        { description, completed },
        {
          headers: {
            Authorization:
              "Bearer " + getState().auth.user.token
          }
        }
      )
      .then(
        ({
          data: { description, completed, _id }
        }: {
          data: TaskType;
        }) => {
          dispatch(
            addTask({
              description,
              completed,
              _id
            })
          );
        }
      )
      .catch();
  };
};

export const startRemoveTask = (id: string) => {
  return (dispatch: any, getState: any) => {
    return client
      .delete(`/tasks/${id}`, {
        headers: {
          Authorization:
            "Bearer " + getState().auth.user.token
        }
      })
      .then((res) => {
        dispatch(removeTask(res.data._id));
      })
      .catch();
  };
};

export const startEditTask = (
  id: string,
  updates: Partial<TaskType>
) => {
  return (dispatch: any, getState: any) => {
    return client
      .patch(`/tasks/${id}`, updates, {
        headers: {
          Authorization:
            "Bearer " + getState().auth.user.token
        }
      })
      .then(() => dispatch(editTask(id, updates)))
      .catch();
  };
};

export const setTasks = (
  tasks: TaskType[]
): SetTasksAction => ({
  type: "SET_TASKS",
  tasks
});

export const addTask = (task: TaskType): AddTaskAction => ({
  type: "ADD_TASK",
  task
});

export const removeTask = (
  id: string
): RemoveTaskAction => ({
  type: "REMOVE_TASK",
  id
});

export const clearTasks = (): ClearTasksAction => ({
  type: "CLEAR_TASKS"
});

export const editTask = (
  id: string,
  updates: Partial<TaskType>
): EditTaskAction => ({
  type: "EDIT_TASK",
  id,
  updates
});

export type TaskActionType =
  | SetTasksAction
  | AddTaskAction
  | RemoveTaskAction
  | ClearTasksAction
  | InitAction
  | EditTaskAction;
