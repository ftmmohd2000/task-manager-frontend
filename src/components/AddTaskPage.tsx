import React from "react";
import TaskForm from "./TaskForm";
import { RouteComponentProps } from "react-router";
import { ConnectedProps, connect } from "react-redux";
import { ThunkDispatch } from "redux-thunk";
import { AnyAction } from "redux";
import { startAddTask } from "../actions/task";

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, AnyAction>
) => ({
  addTask: (description: string, completed: boolean) =>
    dispatch(startAddTask(description, completed))
});

const connector = connect(undefined, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type IAddTaskPageProps = RouteComponentProps &
  PropsFromRedux;

export const AddTaskPage = ({
  history,
  addTask
}: IAddTaskPageProps) => {
  return (
    <div>
      <TaskForm history={history} saveTask={addTask} />
    </div>
  );
};

export default connector(AddTaskPage);
