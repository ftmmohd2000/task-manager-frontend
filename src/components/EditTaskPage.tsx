import React from "react";
import { connect, ConnectedProps } from "react-redux";
import {
  Redirect,
  RouteComponentProps
} from "react-router";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { startEditTask } from "../actions/task";
import { ReduxStateType } from "../store/configureStore";
import TaskType from "../types/task";
import TaskForm from "./TaskForm";

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, AnyAction>
) => ({
  editTask: (id: string, updates: Partial<TaskType>) =>
    dispatch(startEditTask(id, updates))
});

const mapStateToProps = (
  state: ReduxStateType,
  props: RouteComponentProps<{ id: string }>
) => ({
  task: state.task.tasks.find(
    (task) => task._id === props.match.params.id
  ),
  id: props.match.params.id
});

const connector = connect(
  mapStateToProps,
  mapDispatchToProps
);

type PropsFromRedux = ConnectedProps<typeof connector>;

type IEditTaskPageProps = RouteComponentProps<{
  id: string;
}> &
  PropsFromRedux;

export const EditTaskPage = ({
  task,
  history,
  id,
  editTask
}: IEditTaskPageProps) => {
  const startEditTask = (
    description: string,
    completed: boolean
  ) => {
    return editTask(id, {
      description,
      completed
    });
  };

  return task ? (
    <div>
      <TaskForm
        history={history}
        task={task}
        saveTask={startEditTask}
      />
    </div>
  ) : (
    <Redirect to="/dashboard" />
  );
};

export default connector(EditTaskPage);
