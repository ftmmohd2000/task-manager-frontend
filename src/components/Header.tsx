import React from "react";
import { connect, ConnectedProps } from "react-redux";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { startDelete, startLogout } from "../actions/auth";

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, AnyAction>
) => ({
  logout: () => dispatch(startLogout()),
  deleteUser: () => dispatch(startDelete())
});

const connector = connect(undefined, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type IHeaderProps = PropsFromRedux & {
  history: {
    push: (arg0: string) => void;
  };
};

export const Header = ({
  logout,
  deleteUser,
  history
}: IHeaderProps) => {
  const tryLogout = () => {
    logout();
    history.push("/");
  };

  const tryDeleteUser = () => {
    deleteUser();
    history.push("/");
  };

  return (
    <div>
      <h1>Task Manager</h1>
      <button id="logout" onClick={tryLogout}>
        Logout
      </button>
      <button id="deleteUser" onClick={tryDeleteUser}>
        Delete Account
      </button>
    </div>
  );
};

export default connector(Header);
