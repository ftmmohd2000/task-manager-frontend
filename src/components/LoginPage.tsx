import React, { useState } from "react";
import { connect, ConnectedProps } from "react-redux";
import { RouteComponentProps } from "react-router";
import { Link } from "react-router-dom";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { startLogin } from "../actions/auth";
import { startSetTasks } from "../actions/task";

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, AnyAction>) => ({
  setTasks: () => dispatch(startSetTasks()),
  login: (email: string, password: string) =>
    dispatch(startLogin(email, password))
});

const connector = connect(undefined, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux & RouteComponentProps;

const LoginPage = ({ history, setTasks, login }: Props) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");

  const tryLogin = (e: any) => {
    e.preventDefault();
    login(email, password)
      .then(() => {
        setTasks();
        history.push("/dashboard");
      })
      .catch(() => {
        setError("Invalid Credentials");
      });
  };

  return (
    <div>
      <h1>Login Page</h1>
      <form onSubmit={tryLogin}>
        <label>email:</label>
        <input
          type="text"
          value={email}
          onChange={(e: any) => {
            setEmail(e.target.value);
          }}
        />
        <label>password:</label>
        <input
          type="password"
          value={password}
          onChange={(e: any) => {
            setPassword(e.target.value);
          }}
        />
        <button type="submit">Login</button>
      </form>
      {error && <p>{error}</p>}
      <Link to="/signup">Signup</Link>
    </div>
  );
};

export default connector(LoginPage);
