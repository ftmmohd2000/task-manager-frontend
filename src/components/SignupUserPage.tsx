import React, { useState } from "react";
import { RouteComponentProps } from "react-router";
import { startSignup } from "../actions/auth";
import { connect } from "react-redux";
import { ThunkDispatch } from "redux-thunk";
import { AnyAction } from "redux";

interface ISignupUserPageProps extends RouteComponentProps {
  signup: (
    arg0: string,
    arg1: string,
    arg2: string,
    arg3: number
  ) => Promise<any>;
}
export const SignupUserPage = (
  props: ISignupUserPageProps
) => {
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const [age, setAge] = useState(0);

  const tryLogin = (e: any) => {
    e.preventDefault();
    props
      .signup(name, email, password, age)
      .then(() => {
        props.history.push("/dashboard");
      })
      .catch(() => {
        setError("Invalid Credentials");
      });
  };

  return (
    <div>
      <h1>Signup Page</h1>
      <form onSubmit={tryLogin}>
        <p>email: </p>
        <input
          type="text"
          value={email}
          onChange={(e: any) => {
            setEmail(e.target.value);
          }}
        />
        <p>password: </p>
        <input
          type="password"
          value={password}
          onChange={(e: any) => {
            setPassword(e.target.value);
          }}
        />
        <p>name: </p>
        <input
          type="text"
          value={name}
          onChange={(e: any) => {
            setName(e.target.value);
          }}
        />
        <p>age: </p>
        <input
          type="number"
          value={age}
          onChange={(e: any) => {
            setAge(e.target.value);
          }}
        />
        <button>SignUp</button>
      </form>
      {error && <p>{error}</p>}
    </div>
  );
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, AnyAction>
) => ({
  signup: (
    name: string,
    email: string,
    password: string,
    age: number = 0
  ) => dispatch(startSignup(name, email, password, age))
});
export default connect(
  undefined,
  mapDispatchToProps
)(SignupUserPage);
