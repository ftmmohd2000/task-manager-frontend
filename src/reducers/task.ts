import TaskType from "../types/task";
import { TaskActionType } from "../actions/task";

export interface TaskStateType {
  tasks: TaskType[];
}

export const tasksDefaultState = {
  tasks: []
};

export default (
  state: TaskStateType = tasksDefaultState,
  action: TaskActionType
) => {
  switch (action.type) {
    case "SET_TASKS":
      return {
        tasks: action.tasks
      };
    case "ADD_TASK":
      return {
        tasks: state.tasks.concat(action.task)
      };
    case "REMOVE_TASK":
      return {
        tasks: state.tasks.filter(
          (task) => task._id !== action.id
        )
      };
    case "CLEAR_TASKS":
      return {
        tasks: []
      };
    case "EDIT_TASK":
      return {
        tasks: state.tasks.map((task) => {
          if (task._id === action.id)
            return {
              ...task,
              ...action.updates
            };
          else return task;
        })
      };
    default:
      return state;
  }
};
