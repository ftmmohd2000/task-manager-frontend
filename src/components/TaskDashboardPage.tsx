import React from "react";
import { RouteComponentProps } from "react-router";
import { connect, ConnectedProps } from "react-redux";
import { Link } from "react-router-dom";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { startRemoveTask } from "../actions/task";
import Header from "./Header";
import TaskList from "./TaskList";
import TaskType from "../types/task";

const mapStateToProps = (state: any) => ({
  tasks: state.task.tasks
});

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, AnyAction>) => ({
  removeTask: (id: string) => dispatch(startRemoveTask(id))
});

const connector = connect(mapStateToProps, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = RouteComponentProps &
  PropsFromRedux & {
    tasks: TaskType[];
  };

const TaskDashboardPage = ({ removeTask, history, tasks }: Props) => {
  const handleRemove = (id: string) => {
    removeTask(id);
  };

  return (
    <div>
      <Header history={history} />
      <TaskList tasks={tasks} handleRemove={handleRemove} />
      <Link to="/add">Add Task</Link>
    </div>
  );
};

export default connector(TaskDashboardPage);
