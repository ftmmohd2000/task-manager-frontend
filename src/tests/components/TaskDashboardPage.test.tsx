import React from "react";
import { shallow } from "enzyme";
import { TaskDashboardPage } from "../../components/TaskDashboardPage";
import { tasks } from "../fixtures/task";

let wrapper: any,
  removeTask: (argo: string) => Promise<void>;

beforeEach(() => {
  const routeComponentPropsMock = {
    history: { push: jest.fn() } as any,
    match: {} as any,
    location: {} as any
  };
  removeTask = jest.fn(() => Promise.resolve());
  wrapper = shallow(
    <TaskDashboardPage
      {...routeComponentPropsMock}
      removeTask={removeTask}
      tasks={tasks}
    />
  );
});

test("should render TaskDashboardPage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});
