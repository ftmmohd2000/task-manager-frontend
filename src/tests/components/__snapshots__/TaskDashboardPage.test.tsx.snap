// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TaskDashboardPage correctly 1`] = `
<div>
  <Connect(Header)
    history={
      Object {
        "push": [MockFunction],
      }
    }
  />
  <TaskList
    handleRemove={[Function]}
    tasks={
      Array [
        Object {
          "_id": "jihugytdrhgxcvbhjkuyyri76uy",
          "completed": false,
          "description": "Get Swole",
        },
        Object {
          "_id": "jihugyjhuyo8767563teyri76uydh",
          "completed": true,
          "description": "Get groceries",
        },
        Object {
          "_id": "jihugytdrh657tuyg3rheduyyri76uy",
          "completed": false,
          "description": "Im not creative",
        },
      ]
    }
  />
  <Link
    to="/add"
  >
    Add Task
  </Link>
</div>
`;
